### PR TITLE
Verify GPU kernel by cloning a copy for CPU and comparing results

### PR DESCRIPTION
**Pull request by pcf000**
Cleanups and clone-validation-related refactoring.
Working version of '--verifier clone', using the kernel clones from '-host-pipeline partition'.  Clones the root function and reverts async.launch to func.call.
Add a '--verifier clone' run to the resnet50_blk.e2e.mlir test.
Clean up some warnings.

There are a lot of cleanup/refactoring changes.  In particular, since for Option, operator=() calls setValue(), I changed setValue() to "=" for aesthetics.  I also abstracted some repeated sections into functions.

For the actual clone-verification, I take advantage of the pattern of a function calling kernels, where the kernels have arch-specific alternatives in sub-modules.  The caller has async.launch to call the kernels, and a later pass will do the magic to choose the right GPU implementation.  Here, I clone the caller and make it just call the original MLIR, which naturally goes to the CPU.

While I'll try to generalise it to create the caller when given just a kernel, the whole idea does depend on the cloned kernels in sub-modules;  nothing here can cause some code to be compiled for GPU and some for CPU.

### CODE REVIEW

File: mlir/test/rocmlir-driver/populate_host_print.mlir [hunk 1]
```diff
@@ -65,6 +65,8 @@
 // CHECK-NEXT: memref.store {{.*}}[%[[n]], %[[g]], %[[k]], %[[ho]], %[[wo]]] : memref<[[N]]x[[G]]x[[K]]x[[HO]]x[[WO]]x[[TYPE]]>
 
 // CHECK: call @{{.*}}({{.*}}, {{.*}}, {{.*}}) : (memref<[[G]]x[[K]]x[[C]]x[[Y]]x[[X]]x[[TYPE]]>, memref<[[N]]x[[G]]x[[C]]x[[HI]]x[[WI]]x[[TYPE]]>, memref<[[N]]x[[G]]x[[K]]x[[HO]]x[[WO]]x[[TYPE]]>) -> ()
+// CHECK-NEXT: memref.cast %{{.*}} : memref<[[N]]x[[G]]x[[K]]x[[HO]]x[[WO]]x[[TYPE]]> to memref<*x[[TYPE]]>
+// CHECK-NEXT: call @printMemrefF32(%{{.*}}) : (memref<*x[[TYPE]]>) -> ()
```
Comment (pcf000) [inline mlir/test/rocmlir-driver/populate_host_print.mlir:69]: The memref.dealloc calls moved.

File: mlir/test/xmir/e2e/resnet50_blk.e2e.mlir [hunk 2]
```diff
@@ -1,9 +1,18 @@
...
+// RUN: rocmlir-driver -host-pipeline partition,highlevel -targets gfx908,gfx90a %s | rocmlir-gen -ph -print-results -rand_type float -rand 1 -fut resnet50 --verifier clone - | rocmlir-driver -host-pipeline xmodel -kernel-pipeline full -triple amdgcn-amd-amdhsa -targets gfx908,gfx90a | xmir-runner --shared-libs=%linalg_test_lib_dir/libmlir_rocm_runtime%shlibext,%conv_validation_wrapper_library_dir/libconv-validation-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext,%linalg_test_lib_dir/libmlir_c_runner_utils%shlibext --entry-point-result=void | FileCheck %s --check-prefix=CLONE
 
 module {
 // CHECK: Unranked Memref base@ = 0x{{.*}} rank = 4 offset = 0 sizes = [1, 32, 32, 64] strides = [65536, 2048, 64, 1] data =
 // CHECK-NEXT: 0.680375,     -0.211234,     0.566198,     6.59688,     0.823295,     -0.604897,     5.67045,     6.53646,     5.55555,     6.10794,
 
+// CLONE: Number of elements: 65536
```
Comment (pcf000) [inline mlir/test/xmir/e2e/resnet50_blk.e2e.mlir:8]: This is the output of the verifier function.  Not checking for exact diff values, or "[1 1 1]," until we have perfect fidelity.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 3]
```diff
@@ -429,65 +432,50 @@ static void correctParameters() {
...
-    if (paddingHeightLeft.getValue() == 0 &&
-        paddingHeightRight.getValue() == 0) {
-      paddingHeightLeft.setValue(paddingHeight.getValue());
-      paddingHeightRight.setValue(paddingHeight.getValue());
-    } else {
-      if (paddingHeightLeft.getValue() != paddingHeight.getValue() ||
-          paddingHeightRight.getValue() != paddingHeight.getValue()) {
-        llvm::errs()
-            << "you can't use both padding_h and (padding_h_l,padding_h_r).\n";
+  // yxcgk not implement yet
```
Comment (pcf000) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:435]: Cleanup and abstraction.  Checking for length-4 strings is redundant.  The padding checks are the same for both groups.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 4]
```diff
@@ -1736,11 +1715,13 @@ static func::FuncOp createVerifierFunc(ModuleOp &module, const KernelIF &kernel,
           b.create<memref::StoreOp>(loc, valExt, arg1, ivs);
         });
   } else {
-    testResult = b.create<memref::CastOp>(loc, mr5DUnkType, arg0);
+    testResult = makeNDMemRef(b, arg0, 5);
```
Comment (pcf000) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:1718]: I forget exactly why I needed this.  I think I had a test case at a slightly different point in the pipeline.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 5]
```diff
@@ -1831,12 +1812,140 @@ static LogicalResult populateTensorFillLogic(mlir::OpBuilder &b,
   return success();
 }
 
+// Convert the async.launch/async.await pattern back to func.call.
+void undoAsyncLaunchPass(Operation *cloneFunc) {
```
Comment (pcf000) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:1816]: New function.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 6]
```diff
@@ -1831,12 +1812,140 @@ static LogicalResult populateTensorFillLogic(mlir::OpBuilder &b,
...
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  };
+  while (cloneFunc->walk(walker).wasInterrupted()) {
+  }
+}
+
+static void
+insertValidationCalls(const rock::Conv2dGenerator::Config &genConfig,
```
Comment (pcf000) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:1843]: Abstracted from the "if (hasValidation)" section in the original, with "clone" section added.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 7]
```diff
@@ -1971,121 +2081,33 @@ populateHostHarnessLogic(ModuleOp &module,
...
+  if (hasValidation && !hasCloneValidation)
+    insertValidationCalls(genConfig, b, module, valVars, localVars, outIdx,
+                          func, root0);
 
   // Print and cleanup validation vars
   for (auto &vvar : valVars) {
     // print vvar
     if ((vvar == valVars[outIdx]) && printValidationResults.getValue())
       emitPrintTensor(b, vvar);
-    // dealloc vvar
```
Comment (pcf000) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:2087]: Moved all deallocations to the end, to accommodate tests that have more than one root, all of which use the same buffers.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 8]
```diff
@@ -2096,12 +2118,50 @@ populateHostHarnessLogic(ModuleOp &module,
...
+    b.create<memref::DeallocOp>(loc, vvar);
+  }
+  for (auto &lvar : localVars) {
     b.create<memref::DeallocOp>(loc, lvar);
   }
 
   b.create<func::ReturnOp>(loc, ValueRange{});
 
+  // Wrap the kernels and gather them to substitute in calls.
+  llvm::SmallDenseMap<func::FuncOp, func::FuncOp> wrappedFuncs;
```
Comment (pcf000) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:2117]: I could swear I'd already pushed this, but I can't find the pull request.  The commit was from April, though.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 9]
```diff
@@ -1831,12 +1812,140 @@ static LogicalResult populateTensorFillLogic(mlir::OpBuilder &b,
...
+
+      // Remove any token operands, at any position.
+      auto operands = launch->getOperands();
+      BitVector indicesToErase(launch.getNumOperands());
+      for (auto idx : llvm::seq<int>(0, launch.getNumOperands()))
+        if (!launch->getOperand(idx) ||
+            launch->getOperand(idx).getType() == tokenType)
+          indicesToErase.set(idx);
+      launch->eraseOperands(indicesToErase);
+      operands = launch->getOperands();
```
Comment (sjw36) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: You can use 'launch.getCallOperands();' which drops the tokens.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 10]
```diff
@@ -2250,25 +2314,34 @@ int main(int argc, char **argv) {
...
   // Make KernelIFs for the roots, to pass to populateHostHarnessLogic().
   SmallVector<KernelIF, 8> rootIFs;
-  for (auto *node : roots) {
-    func::FuncOp func =
-        dyn_cast<func::FuncOp>(node->getCallableRegion()->getParentOp());
-    rootIFs.emplace_back(func);
-  }
 
   if (testFuncNameVal.empty()) {
+    // Compute set of call-graph root nodes;  they're the ones we need to
```
Comment (pcf000) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:2304]: Roots really only needed when we don't have -fut.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 11]
```diff
@@ -564,64 +552,64 @@ static void populateDefaults() {
...
+      arch = "amdgcn-amd-amdhsa:gfx908";
     }
   }
 
-  if (xdlopsV2.getValue() == true) {
-    num_cu.setValue(120);
-    arch.setValue("amdgcn-amd-amdhsa:gfx908");
+  if (xdlopsV2.getValue()) {
+    num_cu = 120;
+    arch = "amdgcn-amd-amdhsa:gfx908";
```
Comment (krzysz00) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:599]: Come to think it, these 'arch` strings aren't meant to have the triple in them

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp (hunk 5 above)
Comment (krzysz00) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp:1816]: Wouldn't it be better to actually use Simon's PR that gives CPU-side support for `async.launch`?

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp (hunk 4 above)
Comment (krzysz00) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp:1718]: Does this allocate a new memref?

Comment (pcf000) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp:1718]: Despite the name, the only ops that makeNDMemRef() makes are ExpandShape and CollapseShape.  It just adjusts the rank of the reference;  I believe my problem was that CastOp wasn't enough to make it work for my side cases.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp (hunk 5 above)
Comment (pcf000) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp:1816]: Parallel simultaneous evolution.  I'll look at the other pull request and possibly adopt that.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 12]
```diff
@@ -1831,12 +1812,140 @@ static LogicalResult populateTensorFillLogic(mlir::OpBuilder &b,
...
+    // needed.
+    auto *cloneFunc = func->clone();
+    undoAsyncLaunchPass(cloneFunc);
+    SymbolOpInterface cloneFuncOp = dyn_cast<SymbolOpInterface>(cloneFunc);
+    SmallString<128> nameBuffer(cloneFuncOp.getName());
+    nameBuffer += "_cloned";
+    cloneFuncOp.setName(nameBuffer);
+    cloneFunc->removeAttr("kernel");
+    //     cloneFunc->setAttr("original_func", SymbolRefAttr::get(func));
+    //     func->setAttr("clone_func", SymbolRefAttr::get(cloneFunc));
```
Comment (sjw36) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: remove commented code.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 13]
```diff
@@ -2096,12 +2118,50 @@ populateHostHarnessLogic(ModuleOp &module,
...
+  for (auto &kernel : kernels) {
+    if (kernel.func->hasAttr("kernel")) {
+      wrappedFuncs[kernel.func] = createGPUWrapper(module, kernel);
+    } else {
+      wrappedFuncs[kernel.func] = kernel.func;
+    }
+  }
+
+  // Redirect calls to kernel functions to point at wrapped functions.
+  module.walk([&](Operation *op) -> WalkResult {
```
Comment (sjw36) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: Can be run on a CallOpInterface for both CallOp and async::LaunchOp

**Review by sjw36 — APPROVED** [review]
Looks good overall. Thanks.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp (hunk 9 above)
Comment (pcf000) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: That simplified things a lot.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp (hunk 13 above)
Comment (pcf000) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: Also simplified things.

**Pull request merged**